projects/noloan/app/src/main/java/hasadna/noloan/DbMessages.java

```diff
@@ -22,69 +22,110 @@ public static DbMessages getInstance() {
-    return spam;
+  public List<SmsMessage> getMessages() {
+    return messages;
   }
 
-  public List<SmsMessage> getSuggestions() {
-    return suggestions;
-  }
-
-  public void addSpam(SmsMessage spam) {
-    this.spam.add(spam);
+  // Checks if message had already been suggested. Updates counter / Adds a new suggestion
+  public void suggestMessage(SmsMessage sms) {
+
+    // Search if message was already suggested -> Modify counter (+1)
+    SmsMessage message = searchSuggestions(sms.getBody(), sms.getSender());
+    if (message != null) {
+      int index = messages.indexOf(message);
+      SmsMessage newMessage = message.toBuilder().setCounter(message.getCounter() + 1).build();
+      firestoreClient.modifyMessage(message, newMessage);
+      modifyMessage(newMessage);
+      if (messagesListener != null) {
+        messagesListener.messageModified(index);
+      }
+    }
 
-    if (spamListener != null) {
-      spamListener.messageAdded();
+    // Add new suggestion
+    else {
+      firestoreClient.writeMessage(sms.toBuilder().setCounter(sms.getCounter() + 1).build());
     }
   }
 
-  public void spamRemove(SmsMessage spam) {
-    int index = this.spam.indexOf(spam);
-    this.spam.remove(spam);
+  public void undoSuggestion(SmsMessage sms) {
 
-    if (spamListener != null) {
-      spamListener.messageRemoved(index);
+    // If other people suggested this message as well - Modify message with a new counter (-1)
+    if (sms.getCounter() > 1) {
+      SmsMessage newMessage = sms.toBuilder().setCounter(sms.getCounter() - 1).build();
+      firestoreClient.modifyMessage(sms, newMessage);
+      modifyMessage(newMessage);
+      if (messagesListener != null) {
+        messagesListener.messageModified(messages.indexOf(sms));
+      }
     }
-  }
 
-  public void addSuggestion(SmsMessage suggestedSpam) {
-    suggestions.add(suggestedSpam);
+    // In case the user is the only one suggested this spam - Remove suggestion
+    else {
+      removeMessage(sms);
+    }
+  }
 
-    if (suggestionsListener != null) {
-      suggestionsListener.messageAdded();
+  public SmsMessage searchSuggestions(String body, String sender) {
+    SmsMessage sms = null;
+    for (SmsMessage message : messages) {
+      if (message.getBody().contentEquals(body) && message.getSender().contentEquals(sender)) {
+        sms = message;
+      }
     }
+
+    return sms;
   }
 
-  public void suggestionRemove(SmsMessage suggestedSpam) {
-    int index = suggestions.indexOf(suggestedSpam);
-    suggestions.remove(suggestedSpam);
+  // region Functions used by db Listeners when list changes
+  public void addMessage(SmsMessage smsMessage) {
+    messages.add(smsMessage);
+    if (messagesListener != null) {
+      messagesListener.messageAdded();
+    }
+  }
 
-    if (suggestionsListener != null) {
-      suggestionsListener.messageRemoved(index);
+  public void removeMessage(SmsMessage smsMessage) {
+    int index = messages.indexOf(smsMessage);
```

Comment:
Maybe use `getIndexById()`?

Reply:
I changed it. Though I can't say I don't feel "safer" to relay on `List.indexOf(Object)` rather than by id at least up until we straighten a line with the messages in DB.

Reply:
This is actually safer I think, because if a count was increased, you won't find the index based on the `SmsMessage` object, but you will based on the `id`.